### PR TITLE
Device-fied Toaplan GXC MCU and its interface emulation

### PR DESCRIPTION
**Pull request by cam900**
twincobr.cpp: Fix fsharkbt extra MCU behavior with side effects, Fix initializing values, Use shorter/correct type values
toaplan1.cpp: Fix initializing values, Use shorter/correct type values
tms32010.cpp: Prepare support for Internal ROM/Microcomputer mode, Internal IO (TMS32014, TMS32017, custom, etc), Fix 64K program space, Add/Fix notes

### CODE REVIEW

**Comment by Quench0** [comment]
The contribution Darren Olafson and Stephane Humbert made to the Toaplan1 driver are unrelated to the TMS32010 DSP so they should be removed as copyright holders from the GXL DSP code.

**Comment by cam900** [comment]
> The contribution Darren Olafson and Stephane Humbert made to the Toaplan1 driver are unrelated to the TMS32010 DSP so they should be removed as copyright holders from the GXL DSP code.

Thanks, This issue is addressed now.

File: src/mame/drivers/twincobr.cpp [hunk 1]
```diff
@@ -1415,13 +1397,21 @@ void twincobr_state::init_twincobr()
 }
 
 
+void twincobr_state::init_fsharkbt()
+{
+	init_twincobr();
+	m_gxl->dsp().space(AS_IO).install_readwrite_handler(0x02, 0x02,
+		read16smo_delegate(*this, FUNC(twincobr_state::fsharkbt_dsp_r)), write16smo_delegate(*this, FUNC(twincobr_state::fsharkbt_dsp_w)), 0xffff);
+}
```
Comment (cuavas) [inline src/mame/drivers/twincobr.cpp:1405]: I really don’t like the idea of a change that requires adding init functions for installing handlers.  That seems regressive.

File: src/mame/includes/toaplan1.h [hunk 2]
```diff
@@ -11,6 +11,7 @@
 
 #include "cpu/m68000/m68000.h"
 #include "cpu/tms32010/tms32010.h"
+#include "machine/toaplan_gxl.h"
 #include "sound/ymopl.h"
 #include "video/toaplan_scu.h"
```
Comment (cuavas) [inline src/mame/includes/toaplan1.h:16]: Headers from src/mame shouldn’t be intermixed with headers from src/devices.  Headers from src/mame should come first (most dependent to least dependent module).

File: src/mame/includes/twincobr.h [hunk 3]
```diff
@@ -11,6 +11,7 @@
 
 #include "cpu/tms32010/tms32010.h"
 #include "machine/74259.h"
+#include "machine/toaplan_gxl.h"
 #include "video/mc6845.h"
 #include "video/bufsprite.h"
 #include "video/toaplan_scu.h"
```
Comment (cuavas) [inline src/mame/includes/twincobr.h:17]: Same here, headers should be organised by module, and alphebetised within a module.

File: src/mame/machine/toaplan_gxl.h [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+#define MAME_MACHINE_TOAPLAN_GXL_H
+
+#pragma once
+
+#include "cpu/tms32010/tms32010.h"
+
+typedef device_delegate<void (u32 &main_ram_seg, u32 &dsp_addr, u16 data)> dsp_addr_cb_delegate;
+typedef device_delegate<bool (u32 main_ram_seg, u32 dsp_addr, u16 &data)> dsp_read_cb_delegate;
+typedef device_delegate<bool (u32 main_ram_seg, u32 dsp_addr, bool &dsp_execute, u16 data)> dsp_write_cb_delegate;
+
```
Comment (cuavas) [inline src/mame/machine/toaplan_gxl.h:13]: These shouldn’t be in the global namespace – they can be scoped inside the device class.

**Comment by cam900** [comment]
This PR is not match with current source code structure, so close this.

**Pull request closed without merging**